cpp/include/patterns/copy_v_transform_reduce_in_out_nbr.cuh

```diff
@@ -416,60 +360,136 @@ void copy_v_transform_reduce_nbr(raft::handle_t const& handle,
+            <<<update_grid.num_blocks, update_grid.block_size, 0, handle.get_stream()>>>(
+              matrix_partition,
+              matrix_partition.get_major_first(),
+              matrix_partition.get_major_first() + segment_offsets[1],
+              adj_matrix_row_value_input_first + row_value_input_offset,
+              adj_matrix_col_value_input_first + col_value_input_offset,
+              vertex_value_output_first,
+              e_op,
+              major_init);
+        }
+      }
+      if (segment_offsets[2] - segment_offsets[1] > 0) {
+        raft::grid_1d_warp_t update_grid(segment_offsets[2] - segment_offsets[1],
                                          detail::copy_v_transform_reduce_nbr_for_all_block_size,
                                          handle.get_device_properties().maxGridSize[0]);
-
-      if (GraphViewType::is_multi_gpu) {
-        auto& row_comm = handle.get_subcomm(cugraph::partition_2d::key_naming_t().row_name());
-        auto const row_comm_size = row_comm.get_size();
-        auto& col_comm = handle.get_subcomm(cugraph::partition_2d::key_naming_t().col_name());
-        auto const col_comm_rank = col_comm.get_rank();
-
-        auto row_value_input_offset = GraphViewType::is_adj_matrix_transposed
-                                        ? vertex_t{0}
-                                        : matrix_partition.get_major_value_start_offset();
-        auto col_value_input_offset = GraphViewType::is_adj_matrix_transposed
-                                        ? matrix_partition.get_major_value_start_offset()
-                                        : vertex_t{0};
-
-        detail::for_all_major_for_all_nbr_low_degree<in == GraphViewType::is_adj_matrix_transposed>
-          <<<update_grid.num_blocks, update_grid.block_size, 0, handle.get_stream()>>>(
-            matrix_partition,
-            graph_view.get_vertex_partition_first(comm_root_rank),
-            graph_view.get_vertex_partition_last(comm_root_rank),
-            adj_matrix_row_value_input_first + row_value_input_offset,
-            adj_matrix_col_value_input_first + col_value_input_offset,
-            (in == GraphViewType::is_adj_matrix_transposed) ? major_buffer_first
-                                                            : minor_buffer_first,
-            e_op,
-            major_init);
-      } else {
-        detail::for_all_major_for_all_nbr_low_degree<in == GraphViewType::is_adj_matrix_transposed>
-          <<<update_grid.num_blocks, update_grid.block_size, 0, handle.get_stream()>>>(
-            matrix_partition,
-            graph_view.get_vertex_partition_first(comm_root_rank),
-            graph_view.get_vertex_partition_last(comm_root_rank),
-            adj_matrix_row_value_input_first,
-            adj_matrix_col_value_input_first,
-            vertex_value_output_first,
-            e_op,
-            major_init);
+        if (GraphViewType::is_multi_gpu) {
```

Comment:
A minor thought for readability.

Many of these 'if (GraphViewType::is_multi_gpu)` code blocks are identical except for where to put the result.

Would this read easier if the if statements assigned the buffer and there was only one call to `detail::for_all_major_for_all_nbr_mid_degree<update_major>`?

e.g.:
```
auto tmp_output_buffer = (GraphViewType::is_multi_gpu) ?
    (update_major ? major_buffer_first + segment_offests[1] : minor_buffer_first) :
    (vertex_value_output_first + (update_major ? segment_offsets[1] : vertex_t{0}));

detail::for_all_major_for_all_nbr_mid_degree<update_major>
  <<<update_grid.num_blocks, update_grid.block_size, 0, handle.get_stream()>>>(
    matrix_partition,
    matrix_partition.get_major_first() + segment_offsets[1],
    matrix_partition.get_major_first() + segment_offsets[2],
    adj_matrix_row_value_input_first + row_value_input_offset,
    adj_matrix_col_value_input_first + col_value_input_offset,
    tmp_output_buffer,
    e_op,
    major_init);
```

Reply:
Thanks for the suggestion, but this requires ```major_buffer_first, minor_buffer_first``` and ```vertex_value_output_first``` to have the same type (e.g. vertex_value_output_first may include something like discard_iterator or permutation iterator); if not this code will not compile.

C++17 supports ```constexpr if``` (we might be switching to C++17 sometime soon) but AFAIK this doesn't work for ternary expressions (e.g. ```auto tmp_output_buffer = constexpr (GraphViewType::is_multi_gpu) ? buffer_a : buffer_b``` will not compile even with C++17).

Let me know if you're aware of the workaround for this problem and or any other concerns.

Reply:
If we switch to C++17, it seems like we might be able to do this with the functor in between.

https://stackoverflow.com/questions/41011900/equivalent-ternary-operator-for-constexpr-if

```
auto get_output_buffer = [](auto buffer0, auto buffer1) {
  if constexpr (GraphViewType::is_multi_gpu) {
   return buffer0;
  }
  else {
    return buffer1;
  }
}
```

I will consider this update once we finally switch to C++17.

Reply:
I added FIXME comments for this.

Reply:
I think the code would be easier to understand and maintain with this change (when we can make it).  I would consider it a rather minor improvement however, I wouldn't prioritize it particularly high.